File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if(n==0 || head ==NULL)
            return head;

        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* startNode = dummy;
        while(n>0)
        {
            startNode = startNode->next;
            n--;
        }

        ListNode* prev = dummy;
        while(startNode->next!=NULL && startNode !=NULL)
        {
            prev = prev->next;
            startNode = startNode->next;
        }


        if(prev->next->next==NULL)
        {
            prev->next = NULL;
        }
        else {
            prev->next = prev->next->next;
        }


        return dummy->next;
    }
};

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */


func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
    dummyNode := &ListNode{Val : 0 , Next: nil}
	
	if list1 == nil && list2 == nil {
		return nil 
	}

	if list1 != nil && list2 == nil {
		return list1
	}

	if list1 == nil && list2 != nil {
		return list2
	}

	ansNode := dummyNode

	for list1 != nil && list2 != nil {
		if list1.Val > list2.Val{
			dummyNode.Next = list2
			dummyNode = dummyNode.Next
			list2 = list2.Next
		} else {
			dummyNode.Next = list1
			dummyNode = dummyNode.Next
			list1 = list1.Next
		}
	}

	if list1!=nil {
		dummyNode.Next = list1
	}
	if list2 != nil {
		dummyNode.Next = list2
	}

	ansNode = ansNode.Next
	return ansNode

}

func mergeKLists(lists []*ListNode) *ListNode {
	if len(lists) ==0 {
		return nil
	}

	if len(lists) ==1 {
		return lists[0]
	}

	list1 := lists[0]
	finalList := &ListNode{}
	for i:=1; i<len(lists); i++{
		finalList = mergeTwoLists(list1 , lists[i])
		list1 = finalList
	}

	return finalList
}
